server/src/main/java/org/opensearch/common/blobstore/AsyncMultiStreamEncryptedBlobContainer.java

```diff
@@ -108,4 +106,69 @@ public InputStreamContainer provideStream(int partNumber) throws IOException {
         }
 
     }
+
+    static class DecryptedReadContext<T, U> extends ReadContext {
+
+        private final U cryptoContext;
+        private final CryptoHandler<T, U> cryptoHandler;
+        private final long fileSize;
+
+        public DecryptedReadContext(ReadContext readContext, CryptoHandler<T, U> cryptoHandler) {
+            super(readContext);
+            this.cryptoHandler = cryptoHandler;
+            try {
+                this.cryptoContext = this.cryptoHandler.loadEncryptionMetadata(null);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            this.fileSize = this.cryptoHandler.estimateDecryptedLength(cryptoContext, readContext.getBlobSize());
+        }
+
+        @Override
+        public long getBlobSize() {
+            return fileSize;
+        }
+
+        @Override
+        public List<InputStreamContainer> getPartStreams() {
+            return super.getPartStreams().stream().map(this::decrpytInputStreamContainer).collect(Collectors.toList());
+        }
+
+        private InputStreamContainer decrpytInputStreamContainer(InputStreamContainer inputStreamContainer) {
```

Comment:
Q: The current APIs within `CryptoHandler` do not directly support transformation of a `List<InputStreamContainer>` which is held in the `ReadContext` and I had to resort to this hack-y approach.
Can we instead add APIs within `CryptoHandler` which a. Translate the offsets b. Decrypt the stream? (b already exists)

Reply:
I am not sure we can do that . `CryptoHandler` will need to adjust the boundaries,  so that it the start and end lies on frame boundaries . And then it will need to adjust the offsets for those adjustments . If you look at `EncryptedBlobContainer#readBlob(String blobName, long position, long length)` , this is the best option we have .

Reply:
Had an offline chat with @vikasvb90. I shared the reasons for the prevailing design and have a path forward for this.